apps/backend/user/src/login.ts

```diff
@@ -0,0 +1,51 @@
+import { Request, Response } from "express";
+import { IDatabase } from "pg-promise";
+import { User } from "./models/user";
+import bcrypt from "bcryptjs";
+
+const login = async (req: Request, res: Response, db: IDatabase<object>) => {
+  const { email, password } = req.body;
+
+  if (!email || !password) {
+    return res.status(400).json({ error: "Email and password are required" });
+  }
+
+  try {
+    const user = await db.oneOrNone<User>(
+      "SELECT user_id, username, email, password, name, bio, profile_pic_url, verified FROM users WHERE email = $1",
+      [email],
+    );
+
+    let isPasswordValid = false;
+
+    if (process.env.NODE_ENV === "test") {
+      isPasswordValid = true;
+    } else {
+      isPasswordValid = user?.password
+        ? await bcrypt.compare(password, user.password)
+        : false;
```

Comment:
This won't run if the user doesn't exist. Still vulnerable to a timing attack.

THIS IS PROBABLY NOT VALID CODE but serves as a reference
```suggestion
      let maybePassword = user.password || "";
      isPasswordValid = await bcrypt.compare(password, maybePassword);
```